### PR TITLE
chore(gatsby): Cache date formatting in lmdb cache

### PR DESCRIPTION
**Pull request by j0sh77**
## Description

This is a version [of a cache change in](https://github.com/gatsbyjs/gatsby/commit/02be7292dd89861c76d6a63609c3fd7be0f01678#diff-f5bfdc523350c7087a2b5e467433a4dc6f40e74f84c6ae3f2e8d98b616fde4d3) `drupal-release` that we want to port over to our main branch. Instead of an in-memory map, we're now using an lmdb cache. 

## Related Issues

[sc-45894]

### CODE REVIEW

**Comment by KyleAMathews** [comment]
BTW, did you benchmark using LMDB vs. in-memory? Generating the date string w/ Moment.js took ~7-10ms in my testing if I remember correctly. So pulling the string from LMDB could/should be cheaper but I went with an in-memory cache as I was 100% sure that'd be faster.

**Comment by wardpeet** [comment]
Problem with in-memory is that memory is retained for a long time. LMDB has lru cache to speed up a lot of the process.

**Comment by KyleAMathews** [comment]
Right. I'm not saying this is wrong. An additional advantage of the lmdb cache is it's shared across workers and persisted across builds. I just don't have a strong mental model yet for how fast it is like I do with an in memory cache.

**Comment by j0sh77** [comment]
> BTW, did you benchmark using LMDB vs. in-memory? Generating the date string w/ Moment.js took ~7-10ms in my testing if I remember correctly. So pulling the string from LMDB could/should be cheaper but I went with an in-memory cache as I was 100% sure that'd be faster.

Good call. Here's the test function I used:
```
import { formatDate } from "./date";
  
const count = BigInt(1000)

function random(min, max) {  
  return Math.floor(Math.random() * (max - min) + min)
}

async function getAverageTiming(getArg: () => any, cb: (arg: any) => void) {
  let totalTime = BigInt(0);
  for (let i = 0; i < count; i++) {
    const arg = getArg()
    const start = process.hrtime.bigint()
    await cb(arg)
    totalTime += process.hrtime.bigint() - start
  }
  return totalTime / count / BigInt(1000)
}

async function test() {
  const formatString = `YYYY-MM-DD HH:mm:ss`;
  const staticDate = new Date(2020, 1, 1)
  
  const staticTime = await getAverageTiming(
    () => ({
      date: staticDate,
      formatString,
    }),
    formatDate
  )

  const dynamicTime = await getAverageTiming(
    () => ({
      date: new Date(random(1970, 2022), random(1, 12), random(1, 28)),
      formatString,
    }),
    formatDate
  )
  
  console.log(`Iterations: ${count}, Static: ${staticTime}μs, Dynamic: ${dynamicTime}μs`)
}

test()
```

And the results are as follows:
```
No cache:
Iterations: 1000, Static: 399μs, Dynamic: 403μs

Cache with in-memory map:
Iterations: 1000, Static: 6μs, Dynamic: 408μs

Cache with lmdb:
Iterations: 1000, Static: 8μs, Dynamic: 401μs
```

File: packages/gatsby/src/schema/types/date.ts [hunk 1]
```diff
@@ -209,19 +210,34 @@ export function isDate(value: MomentInput): boolean {
...
   fromNow,
   difference,
   formatString,
   locale = `en`,
-}: IFormatDateArgs): string | number => {
+}: IFormatDateArgs): Promise<string | number> => {
   const normalizedDate = JSON.parse(JSON.stringify(date))
   if (formatString) {
-    return moment
+    const cacheKey = `${normalizedDate}-${formatString}`
```
Comment (pieh) [inline packages/gatsby/src/schema/types/date.ts:227]: `locale` is also used, we should use it for cache key
```suggestion
    const cacheKey = `${normalizedDate}-${formatString}-${locale}`
```

**Review by pieh — APPROVED** [review]

**Pull request merged**